### PR TITLE
store writer framework ide files - WF-67

### PR DESCRIPTION
**Pull request by FabienArcellier**
* file are removed when page or workflow are removed
* attributes in json line are ordered using id, type, then alphabetical order
* components in files are sorted using their natural order

### CODE REVIEW

File: apps/quickstart/.wf/components-page-0-e8e4m9mx50wfpdty.jsonl [hunk 1]
```diff
@@ -1,13 +1,4 @@
...
-{"id": "yfyr3y9xdllskii4", "type": "sliderinput", "content": {"label": "Cluster deviation", "minValue": "0", "maxValue": "10", "stepSize": "0.1"}, "parentId": "t3zfxwztd4cozu6v", "position": 2, "handlers": {"wf-number-change": "update"}, "binding": {"eventType": "wf-number-change", "stateRef": "cluster_std"}, "visible": {"expression": true, "binding": "", "reversed": false}}
-{"id": "nfkazqjoivq5b4k9", "type": "button", "content": {"text": "Regenerate"}, "parentId": "t3zfxwztd4cozu6v", "position": 4, "handlers": {"wf-click": "update"}, "visible": {"expression": true, "binding": "", "reversed": false}}
-{"id": "7hlk8jwfhvz29rbc", "type": "section", "content": {"title": ""}, "parentId": "e8e4m9mx50wfpdty", "position": 1, "handlers": {}, "visible": {"expression": true, "binding": "", "reversed": false}}
+{"id": "e8e4m9mx50wfpdty", "type": "page", "content": {}, "handlers": {}, "parentId": "root", "position": 0, "visible": {"binding": "", "expression": true, "reversed": false}}
```
Comment (FabienArcellier) [inline apps/quickstart/.wf/components-page-0-e8e4m9mx50wfpdty.jsonl:1]: something wrong here

File: src/writer/core_ui.py [hunk 2]
```diff
@@ -388,32 +388,43 @@ def ingest_bmc_component_tree(component_tree: ComponentTree, components: Dict[st
...
     """
-    component: Optional[ComponentDefinition] = components.get(component_id, None)
-    if component is None:
-        return None
 
-    parent_id = component.get("parentId")
-    if parent_id is None:
-        return None
+    if parent is not None:
+        parent_list = {parent}
```
Comment (FabienArcellier) [inline src/writer/core_ui.py:399]: there a bug here if the ui.json is not sorted (that happens)

File: src/writer/core_ui.py [hunk 3]
```diff
@@ -388,31 +388,34 @@ def ingest_bmc_component_tree(component_tree: ComponentTree, components: Dict[st
...
-        else:
-            return lookup_parent_type_for_component(components, parent_id, parent_type)
 
-    return None
+    if parent is not None:
+        parent_list = {parent}
+        target_components: Dict[str, ComponentDefinition] = {}
+        has_changed = True
+        # in some cases, the order requires to loops few times
+        while True:
```
Comment (FabienArcellier) [inline src/writer/core_ui.py:403]: We can simplify this code if we ensure order of component into audit_and_fix process.

File: src/writer/wf_project.py [hunk 4]
```diff
@@ -141,3 +122,102 @@ def create_default_workflows_root(abs_path: str) -> None:
...
+    list_component_files = set(_list_component_files(wf_directory))
+
+    obsolete_components_files = list_component_files - expected_component_files
+    for f in obsolete_components_files:
+        os.remove(os.path.join(wf_directory, f))
+
+
+def _sort_wf_component_keys(obj: typing.Union[ComponentDefinition, dict]) -> Any:
+    """
+    Sorts the keys of the object recursively to have a consistent order in the json file.
```
Comment (FabienArcellier) [inline src/writer/wf_project.py:171]: write an exemple in documentation

File: src/writer/wf_project.py [hunk 5]
```diff
@@ -141,3 +122,102 @@ def create_default_workflows_root(abs_path: str) -> None:
...
+
+def _write_root_files(wf_directory, components):
+    for root in ROOTS:
+        root_component = components.get(root, None)
+        if root_component:
+            with io.open(os.path.join(wf_directory, f"components-{root}.jsonl"), "w") as f:
+                f.write(json.dumps(_sort_wf_component_keys(root_component)))
+
+
+def _order_components(components: Dict[str, ComponentDefinition]) -> List[ComponentDefinition]:
```
Comment (FabienArcellier) [inline src/writer/wf_project.py:207]: rename into _sort_components

File: tests/backend/test_wf_project.py [hunk 6]
```diff
@@ -97,6 +97,24 @@ def test_wf_project_read_files_should_read_files_in_wf_directory():
     assert isinstance(sc, dict)
     assert len(sc) != 0
 
+
+def test_wf_project_should_remove_obsolete_pages_in_wf_directory():
```
Comment (ramedina86) [inline tests/backend/test_wf_project.py:101]: Nice test case ;)

**Review by ramedina86 — APPROVED** [review]

**Pull request merged**